Lib/test/test_socket.py

```diff
@@ -7058,21 +7061,95 @@ def close_fds(fds):
         self.assertEqual(flags, 0)
         # don't test addr
 
         # test that file descriptors are connected
-        for index, fds in enumerate(pipes):
-            rfd, wfd = fds
-            os.write(wfd, str(index).encode())
+        for index, ((_, wfd), rfd) in enumerate(zip(pipes, fds2)):
+            self._test_pipe(rfd, wfd, str(index).encode())
+
+    @unittest.skipUnless(sys.platform in ("linux", "android", "darwin"),
+                         "works on Linux and macOS")
+    def test_send_recv_fds_with_addrs(self):
+        sock1 = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+        sock2 = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+        rfd, wfd = os.pipe()
+        self.addCleanup(os.close, rfd)
+        self.addCleanup(os.close, wfd)
+
+        with tempfile.TemporaryDirectory() as tmpdir, sock1, sock2:
+            sock1_addr = os.path.join(tmpdir, "sock1")
+            sock2_addr = os.path.join(tmpdir, "sock2")
+            sock1.bind(sock1_addr)
+            sock2.bind(sock2_addr)
+            sock2.setblocking(False)
+
+            socket.send_fds(sock1, [MSG], [rfd], address=sock2_addr)
+            msg, fds, flags, addr = socket.recv_fds(sock2, len(MSG), 1)
+            self._cleanup_fds(fds)
+
+        self.assertEqual(msg, MSG)
+        self.assertEqual(len(fds), 1)
+        self.assertEqual(addr, sock1_addr)
+
+        self._test_pipe(fds[0], wfd, MSG)
 
-        for index, rfd in enumerate(fds2):
-            data = os.read(rfd, 100)
-            self.assertEqual(data,  str(index).encode())
+    @requireAttrs(socket, "MSG_PEEK")
+    @unittest.skipUnless(sys.platform in ("linux", "android"), "works on Linux")
+    def test_recv_fds_peek(self):
+        rfd, wfd = os.pipe()
+        self.addCleanup(os.close, rfd)
+        self.addCleanup(os.close, wfd)
+
+        sock1, sock2 = socket.socketpair(socket.AF_UNIX, socket.SOCK_DGRAM)
+        with sock1, sock2:
+            socket.send_fds(sock1, [MSG], [rfd])
+            sock2.setblocking(False)
+
+            # peek message on sock2
+            peek_len = len(MSG) // 2
+            msg, fds, flags, addr = socket.recv_fds(sock2, peek_len, 1,
+                                                    socket.MSG_PEEK)
+            self._cleanup_fds(fds)
+
+            self.assertEqual(len(msg), peek_len)
+            self.assertEqual(msg, MSG[:peek_len])
+            self.assertEqual(flags & socket.MSG_TRUNC, socket.MSG_TRUNC)
+            self.assertEqual(len(fds), 1)
+            self._test_pipe(fds[0], wfd, MSG)
+
+            # will raise BlockingIOError if MSG_PEEK didn't work
+            msg, fds, flags, addr = socket.recv_fds(sock2, len(MSG), 1)
+            self._cleanup_fds(fds)
+
+            self.assertEqual(msg, MSG)
+            self.assertEqual(len(fds), 1)
+            self._test_pipe(fds[0], wfd, MSG)
+
+    @requireAttrs(socket, "MSG_DONTWAIT")
+    @unittest.skipUnless(sys.platform in ("linux", "android"), "Linux specific test")
```

Comment:
ditto

Reply:
[CI shows](https://github.com/python/cpython/actions/runs/12790887759/job/35657752059?pr=128882) that sendmsg raises `OSError: [Errno 55] No buffer space available` on macOS. I'm not familiar with macOS, so I don't know if it's the expected behavior.